### PR TITLE
Fix flaky test failure in test_cuda_device_memory_allocated

### PR DESCRIPTION
**Pull request by xwang233**
Test test_cuda_device_memory_allocated behaves flaky and fails in our CI test. However, it looks like it's passed in Github CI probably because there are those pytest `--rerun` options.

The test passes if runs by itself:
```python
root@a5bc36a88b05:/opt/pytorch/pytorch# python test/test_cuda_multigpu.py -v -f -k cuda_device_memory_allocated
test_cuda_device_memory_allocated (__main__.TestCudaMultiGPU) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.116s

OK
```

The test fails if runs together with `test_caching_pinned_memory_multi_gpu`:
```python
root@a5bc36a88b05:/opt/pytorch/pytorch# python test/test_cuda_multigpu.py -v -f -k cuda_device_memory_allocated -k caching_pinned_memory
test_caching_pinned_memory_multi_gpu (__main__.TestCudaMultiGPU) ... /opt/pytorch/pytorch/test/test_cuda_multigpu.py:931: UserWarning: The torch.cuda.*DtypeTensor constructors are no longer recommended. It's best to use methods such as torch.tensor(data, dtype=*, device='cuda') to create tensors. (Triggered internally at /opt/pytorch/pytorch/torch/csrc/tensor/python_tensor.cpp:83.)
  gpu_tensor0 = torch.cuda.FloatTensor([0], device=0)
ok
test_cuda_device_memory_allocated (__main__.TestCudaMultiGPU) ... FAIL

======================================================================
FAIL: test_cuda_device_memory_allocated (__main__.TestCudaMultiGPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/pytorch/pytorch/torch/testing/_internal/common_utils.py", line 2362, in wrapper
    method(*args, **kwargs)
  File "/opt/pytorch/pytorch/test/test_cuda_multigpu.py", line 1291, in test_cuda_device_memory_allocated
    self.assertGreater(memory_allocated(0), current_alloc[0])
AssertionError: 512 not greater than 512

To execute this test, run the following from the base repo dir:
     python test/test_cuda_multigpu.py -k test_cuda_device_memory_allocated

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 2 tests in 1.054s

FAILED (failures=1)
```

This fix will ensure it doesn't fail while running along with other tests in the same test module.

Related:
- my wrongful fix in https://github.com/pytorch/pytorch/pull/105501

internal ref: /4208024

### CODE REVIEW

File: test/test_cuda_multigpu.py [hunk 1]
```diff
@@ -945,6 +945,10 @@ def test_caching_pinned_memory_multi_gpu(self):
         self.assertEqual(gpu_tensor1[0], 1)
         self.assertEqual(gpu_tensor0[0], 2)
 
+        # cleanup after `del t`
```
Comment (eqy) [inline test/test_cuda_multigpu.py:948]: Same as the other PR: can we add comments to briefly describe the issue if these are workarounds while we don't know the actual root cause in the caching allocator?

Comment (xwang233) [reply test/test_cuda_multigpu.py:948]: Related: https://github.com/pytorch/pytorch/pull/33575

I would argue if this is a workaround or fix for flaky test. Cuda memory management is not simple task. We should be careful when implementing unit tests to make sure they do not have weird interactions with other units.

Comment (ptrblck) [reply test/test_cuda_multigpu.py:948]: How do we know the failing test isn't uncovering a real issue?
Adding workaround to failing tests won't fix the actual bug, but will increase the technical debt instead.

**Comment by xwang233** [comment]
I investigated a bit deeper into this problem. Here is a short script that reproduces the test error. Note: error may only be reproducible on fast GPUs, such as A100.
```python
import gc

import torch

from torch.testing._internal.common_utils import TestCase
# from unittest import TestCase

class TestA(TestCase):
    def test_a_caching_pinned_memory_multi_gpu(self):
        cycles_per_ms = 1086075.2822836833

        t = torch.FloatTensor([1]).pin_memory(); print('first t', id(t))
        ptr = t.data_ptr()
        gpu_tensor0 = torch.cuda.FloatTensor([0], device=0); print('gpu_tensor0', id(gpu_tensor0))
        gpu_tensor1 = torch.cuda.FloatTensor([0], device=1); print('gpu_tensor1', id(gpu_tensor1))

        with torch.cuda.device(1):
            torch.cuda._sleep(int(1000 * cycles_per_ms))  # delay the copy by 1s
            gpu_tensor1.copy_(t, non_blocking=True)

        del t
        t = torch.FloatTensor([2]).pin_memory(); print('second t', id(t))

        self.assertNotEqual(t.data_ptr(), ptr) ### This line creates massive objects that are not gc'ed fast enough.

        with torch.cuda.device(0):
            gpu_tensor0.copy_(t, non_blocking=True)

        self.assertEqual(gpu_tensor1[0], 1)
        self.assertEqual(gpu_tensor0[0], 2)

    def test_b_cuda_device_memory_allocated(self):
        for o in gc.get_objects():
            if isinstance(o, torch.Tensor):
                print('existing tensor objects', o, id(o))
                # for frame in gc.get_referrers(o):
                #     print(str(frame)[:1024])
    
        from torch.cuda import memory_allocated
        device_count = torch.cuda.device_count()

        old_alloc = [memory_allocated(idx) for idx in range(device_count)]
        print('old_alloc', old_alloc)

        x = torch.ones(10, device="cuda:0"); print('x', id(x))

        new_alloc = [memory_allocated(idx) for idx in range(device_count)]
        print('new_alloc', new_alloc)

        try:
            self.assertGreater(new_alloc[0], old_alloc[0])
            self.assertEqual(new_alloc[1:], old_alloc[1:])
        except:
            for o in gc.get_objects():
                if isinstance(o, torch.Tensor):
                    print('existing tensor objects', o, id(o))
            raise

if __name__ == '__main__':
    import unittest
    unittest.main()
```

Running the script gives the following error message:
```
root@7878ef0b9a5f:/opt/pytorch/pytorch# python a.py -f -v
test_a_caching_pinned_memory_multi_gpu (__main__.TestA) ... first t 140010842580096
/opt/pytorch/pytorch/a.py:14: UserWarning: The torch.cuda.*DtypeTensor constructors are no longer recommended. It's best to use methods such as torch.tensor(data, dtype=*, device='cuda') to create tensors. (Triggered internally at /opt/pytorch/pytorch/torch/csrc/tensor/python_tensor.cpp:83.)
  gpu_tensor0 = torch.cuda.FloatTensor([0], device=0); print('gpu_tensor0', id(gpu_tensor0))
gpu_tensor0 140010842574096
gpu_tensor1 140010522534960
second t 140010522535280
ok
test_b_cuda_device_memory_allocated (__main__.TestA) ... existing tensor objects tensor([2.], device='cuda:0') 140010842574096
existing tensor objects tensor([1.], device='cuda:1') 140010522534960
existing tensor objects tensor([2.]) 140010522535280
/opt/pytorch/pytorch/torch/distributed/distributed_c10d.py:350: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn(
old_alloc [512, 512, 0, 0, 0, 0, 0, 0]
x 140017612144480
new_alloc [512, 0, 0, 0, 0, 0, 0, 0]
existing tensor objects tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.], device='cuda:0') 140017612144480
FAIL

======================================================================
FAIL: test_b_cuda_device_memory_allocated (__main__.TestA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/pytorch/pytorch/torch/testing/_internal/common_utils.py", line 2362, in wrapper
    method(*args, **kwargs)
  File "/opt/pytorch/pytorch/a.py", line 49, in test_b_cuda_device_memory_allocated
    self.assertGreater(new_alloc[0], old_alloc[0])
AssertionError: 512 not greater than 512

To execute this test, run the following from the base repo dir:
     python a.py -k test_b_cuda_device_memory_allocated

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 2 tests in 1.631s

FAILED (failures=1)
```

It is seen that, when the second test `test_b_cuda_device_memory_allocated` started, three tensor objects created in the first test (gpu_tensor0, gpu_tensor1, second t) are still not yet GC'ed and are still occupying GPU memories, thus are being captured by `old_alloc = torch.cuda.memory_allocated()` function call.

Coincidentally, since we don't control when python interpreter starts a GC, when the new tensor `x` is allocated, the GC process happens at the same time in the background and cleans those dangling tensors from previous tests. New `memory_allocated` call then only detected the newly allocated tensor `x`, and thus resulted in unfair comparisons for GPU memory allocations in `test_b`.

I commented on this line
```
self.assertNotEqual(t.data_ptr(), ptr) ### This line creates many objects that are not gc'ed fast enough.
```
If we remove this line from the test script, both tests passed. We can also use `assertEqual` from python native unittest module by changing it here
https://github.com/pytorch/pytorch/blob/01069ad4be449f376cf88a56d842b8eb50f6e9b6/torch/testing/_internal/common_utils.py#L3264-L3267
with 
```
super().assertEqual(x, y)
```

In both ways, the job log is much cleaner:
```
root@7878ef0b9a5f:/opt/pytorch/pytorch# python a.py -f -v
test_a_caching_pinned_memory_multi_gpu (__main__.TestA) ... first t 140692798116208
/opt/pytorch/pytorch/a.py:14: UserWarning: The torch.cuda.*DtypeTensor constructors are no longer recommended. It's best to use methods such as torch.tensor(data, dtype=*, device='cuda') to create tensors. (Triggered internally at /opt/pytorch/pytorch/torch/csrc/tensor/python_tensor.cpp:83.)
  gpu_tensor0 = torch.cuda.FloatTensor([0], device=0); print('gpu_tensor0', id(gpu_tensor0))
gpu_tensor0 140693008161888
gpu_tensor1 140692709220224
second t 140692709221424
ok
test_b_cuda_device_memory_allocated (__main__.TestA) ... /opt/pytorch/pytorch/torch/distributed/distributed_c10d.py:350: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn(
old_alloc [0, 0, 0, 0, 0, 0, 0, 0]
x 140692709221424
new_alloc [512, 0, 0, 0, 0, 0, 0, 0]
ok

----------------------------------------------------------------------
Ran 2 tests in 1.406s

OK
```

I believe that those ErrorMeta exception raises and catches play a role in creating many objects that are not recycled by python interpreter fast enough in the process of exceptions.

This can be verified by commenting out those lines in `test_b`:
```python
    def test_b_cuda_device_memory_allocated(self):
        for o in gc.get_objects():
            if isinstance(o, torch.Tensor):
                print('existing tensor objects', o, id(o))
                # for frame in gc.get_referrers(o):
                #     print(str(frame)[:1024])
```

There are several solutions/workarounds/fixes to this problem:
- Add explicit `gc.collect(); torch.cuda.empty_cache()` after those `assertEqual` function calls which created a lot of python objects that are blocking GPU tensors from being GC'ed by python interpreter.
- Rework assertEqual to create less object references and raises less exceptions.
- Use pytest's reruns just like Github CI does. While the test first fails with non-GC'ed objects, the consequent reruns will make sure the test runs correctly.

**Comment by eqy** [comment]
Sorry if I'm missing something, but is there an explanation for why we still need `empty_cache` after `gc.collect()` IIRC `allocated_bytes.all.current` should not be controlled/dependent on `empty_cache`, which should only affect the _reserved_ bytes?

**Comment by eqy** [comment]
CC @zdevito who may know more details about stats reporting

**Comment by zdevito** [comment]
From reading this code, no tensors should be alive after `test_a_caching_pinned_memory_multi_gpu`. Python will clean up anything that doesn't have a cycle immediately after it goes dead and I do not see a cycle created in this test. So something is wrong with that test. Forcing GC at the end of the test would be covering up whatever is keeping those tensors alive.

**Comment by zdevito** [comment]
It does appear to be the ErrorMeta objects that hold references to these tensors, here is the cycle as found by our `warn_tensor_cycles()` tool in `torch.utils.viz._cycles`.

https://dreampuf.github.io/GraphvizOnline/#digraph%20GraphName%20%7B%0Anode%20%5Bshape%3Drect%5D%3B%0Arankdir%3DLR%3B%0A0%20%5Blabel%3D%22object%5Cntorch.Tensor%22%2C%20color%3Dred%5D%3B%0A1%20%5Blabel%3D%22object%5Cntorch.Tensor%22%2C%20color%3Dred%5D%3B%0A2%20%5Blabel%3D%22%5BErrorMeta%5D%22%2C%20color%3Dblack%5D%3B%0A3%20%5Blabel%3D%22object%5Cntorch.testing._comparison.ErrorMeta%22%2C%20color%3Dblack%5D%3B%0A4%20%5Blabel%3D%22object%5Cnbuiltins.traceback%22%2C%20color%3Dblack%5D%3B%0A5%20%5Blabel%3D%22frame%5Cn...%2Ftorch%2Ftesting%2F_comparison.py%3A363%22%2C%20color%3Dblack%5D%3B%0A6%20%5Blabel%3D%22object%5Cnbuiltins.traceback%22%2C%20color%3Dblack%5D%3B%0A7%20%5Blabel%3D%22frame%5Cn...%2Ftorch%2Ftesting%2F_comparison.py%3A598%22%2C%20color%3Dblack%5D%3B%0A8%20%5Blabel%3D%22object%5Cnbuiltins.traceback%22%2C%20color%3Dblack%5D%3B%0A9%20%5Blabel%3D%22frame%5Cn...%2Ftorch%2Ftesting%2F_comparison.py%3A1238%22%2C%20color%3Dblack%5D%3B%0A10%20%5Blabel%3D%22frame%5Cn...ing%2F_internal%2Fcommon_utils.py%3A3257%22%2C%20color%3Dblack%5D%3B%0A11%20%5Blabel%3D%22frame%5Cn...ing%2F_internal%2Fcommon_utils.py%3A3267%22%2C%20color%3Dblack%5D%3B%0A12%20%5Blabel%3D%22frame%5Cn%2Fraid%2Fzdevito%2Fpytorch%2Fa.py%3A34%22%2C%20color%3Dblack%5D%3B%0A2%20-%3E%203%20%5Blabel%20%3D%20%22%5B0%5D%22%5D%0A3%20-%3E%208%20%5Blabel%20%3D%20%22%3F%22%5D%0A4%20-%3E%205%20%5Blabel%20%3D%20%22%3F%22%5D%0A5%20-%3E%207%20%5Blabel%20%3D%20%22f_back%22%5D%0A6%20-%3E%204%20%5Blabel%20%3D%20%22%3F%22%5D%0A6%20-%3E%207%20%5Blabel%20%3D%20%22%3F%22%5D%0A7%20-%3E%209%20%5Blabel%20%3D%20%22f_back%22%5D%0A8%20-%3E%206%20%5Blabel%20%3D%20%22%3F%22%5D%0A8%20-%3E%209%20%5Blabel%20%3D%20%22%3F%22%5D%0A9%20-%3E%2010%20%5Blabel%20%3D%20%22f_back%22%5D%0A9%20-%3E%202%20%5Blabel%20%3D%20%22local%20error_metas%22%5D%0A10%20-%3E%2011%20%5Blabel%20%3D%20%22f_back%22%5D%0A11%20-%3E%2012%20%5Blabel%20%3D%20%22f_back%22%5D%0A12%20-%3E%200%20%5Blabel%20%3D%20%22local%20gpu_tensor0%22%5D%0A12%20-%3E%201%20%5Blabel%20%3D%20%22local%20gpu_tensor1%22%5D%0A%7D

**Comment by zdevito** [comment]
Here is a fix for the ErrorMeta cycle, it works for me on your example file. Does it also work on your end? https://github.com/pytorch/pytorch/pull/106328

**Comment by xwang233** [comment]
Closing this workaround since it's been fixed by https://github.com/pytorch/pytorch/pull/106328

**Pull request closed without merging**